### PR TITLE
port SpanHelpers.IndexOf(ref byte, byte, int) to Vector128/256

### PR DESCRIPTION
**Pull request by adamsitnik**
## x64

10% improvement for AVX2, no re regressions for AVX and no HI.

<details>

```ini
BenchmarkDotNet=v0.13.1.1828-nightly, OS=Windows 11 (10.0.22000.795/21H2)
AMD Ryzen Threadripper PRO 3945WX 12-Cores, 1 CPU, 24 logical and 12 physical cores
.NET SDK=7.0.100-preview.6.22352.1
  [Host]     : .NET 7.0.0 (7.0.22.32404), X64 RyuJIT AVX2
  Job-BJEYEU : .NET 7.0.0 (42.42.42.42424), X64 RyuJIT AVX2
  Job-KDQZPU : .NET 7.0.0 (42.42.42.42424), X64 RyuJIT AVX2
```

|       Method |        Job |             Toolchain | Size |     Mean | Ratio |
|------------- |----------- |---------------------- |----- |---------:|------:|
| IndexOfValue | Job-BJEYEU |       \PR\corerun.exe |  512 | 6.896 ns |  0.89 |
| IndexOfValue | Job-KDQZPU | \baseline\corerun.exe |  512 | 7.740 ns |  1.00 |

```ini
BenchmarkDotNet=v0.13.1.1828-nightly, OS=Windows 11 (10.0.22000.795/21H2)
AMD Ryzen Threadripper PRO 3945WX 12-Cores, 1 CPU, 24 logical and 12 physical cores
.NET SDK=7.0.100-preview.6.22352.1
  [Host]     : .NET 7.0.0 (7.0.22.32404), X64 RyuJIT AVX2
  Job-SUXHIF : .NET 7.0.0 (42.42.42.42424), X64 RyuJIT AVX
  Job-KVLSYC : .NET 7.0.0 (42.42.42.42424), X64 RyuJIT AVX

EnvironmentVariables=COMPlus_EnableAVX2=0
```

|       Method |        Job |             Toolchain | Size |     Mean | Ratio |
|------------- |----------- |---------------------- |----- |---------:|------:|
| IndexOfValue | Job-JAXEOB |       \PR\corerun.exe |  512 | 9.438 ns |  1.00 |
| IndexOfValue | Job-JLYYUT | \baseline\corerun.exe |  512 | 9.466 ns |  1.00 |

```ini
BenchmarkDotNet=v0.13.1.1828-nightly, OS=Windows 11 (10.0.22000.795/21H2)
AMD Ryzen Threadripper PRO 3945WX 12-Cores, 1 CPU, 24 logical and 12 physical cores
.NET SDK=7.0.100-preview.6.22352.1
  [Host]     : .NET 7.0.0 (7.0.22.32404), X64 RyuJIT AVX2
  Job-KLMBAP : .NET 7.0.0 (42.42.42.42424), X64 RyuJIT
  Job-KSWPBM : .NET 7.0.0 (42.42.42.42424), X64 RyuJIT

EnvironmentVariables=COMPlus_EnableHWIntrinsic=0
```

|       Method |        Job |             Toolchain | Size |     Mean | Ratio |
|------------- |----------- |---------------------- |----- |---------:|------:|
| IndexOfValue | Job-KLMBAP |       \PR\corerun.exe |  512 | 35.78 ns |  1.01 |
| IndexOfValue | Job-KSWPBM | \baseline\corerun.exe |  512 | 35.35 ns |  1.00 |

</details>

## arm64

The initial implementation got x2 perf hit, but after I've moved the call to `ExtractMostSignificantBits` to be performed only when a match is found, the perf is on par (see the second commit). 

BDN reports half a nanosecond difference, which translates to 2-3% and I think that we should just ignore it (it's within the range of error).

<details>

```ini
BenchmarkDotNet=v0.13.1.1828-nightly, OS=ubuntu 20.04
Unknown processor
.NET SDK=7.0.100-rc.1.22403.8
  [Host]     : .NET 7.0.0 (7.0.22.40210), Arm64 RyuJIT AdvSIMD
  Job-SUFTPF : .NET 7.0.0 (42.42.42.42424), Arm64 RyuJIT AdvSIMD
  Job-WFBVQA : .NET 7.0.0 (42.42.42.42424), Arm64 RyuJIT AdvSIMD
```

|       Method |        Job |      Toolchain | Size |     Mean | Ratio |
|------------- |----------- |--------------- |----- |---------:|------:|
| IndexOfValue | Job-SUFTPF |    /PR/corerun |  512 | 20.06 ns |  0.97 |
| IndexOfValue | Job-WFBVQA |  /main/corerun |  512 | 20.78 ns |  1.00 |

</details>

contributes to #64451

### CODE REVIEW

**Review by filipnavara — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs [hunk 1]
```diff
@@ -596,18 +596,19 @@ public static unsafe int IndexOf(ref byte searchSpace, byte value, int length)
...
+                        Vector128<byte> compareResult = Vector128.Equals(values, search);
+                        if (compareResult == Vector128<byte>.Zero)
                         {
                             // Zero flags set so no matches
                             offset += (nuint)Vector128<byte>.Count;
                             continue;
                         }
 
                         // Find bitflag offset of first match and add to current offset
+                        uint matches = compareResult.ExtractMostSignificantBits();
```
Comment (EgorBo) [inline src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs:611]: Last time I tried to do the same it was a noticeable regression for ARM64

Comment (adamsitnik) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs:611]: Same here, but https://github.com/dotnet/runtime/pull/73364/commits/28aa08214b44042c5c14d85bc5512820f775d040 solved the problem

Comment (adamsitnik) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs:611]: Now this operation is performed only once, after we find a match (not once for every vector we compare)

**Review by EgorBo — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs (hunk 1 above)
Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs:611]: What is the cost of this approach vs doing it every loop for `Vector256<T>`

Is it better, particularly for large inputs, to do this there as well?

**Pull request merged**